### DISCUSSION TITLE
Adding custom experiments

### DISCUSSION THREAD

**Discussion by ustojiljkoff** [emdgroup/baybe#104] [Q&A] [answered]
Hi everyone!

Is there a way to add custom experiment results that were never recommended during the campaign? For example, let's say that in the n-th iteration, BayBE proposed the following set of experiments:

```
Recommended experiments: 
      Solvent             Base         Ligand  Temperature  Concentration
107      DMAc  Cesium pivalate  (t-Bu)PhCPhos        120.0          0.153
401  p-Xylene   Cesium acetate  (t-Bu)PhCPhos        105.0          0.153
386  p-Xylene   Cesium acetate      BrettPhos        120.0          0.153
```

These experiments were performed and the results are ready to be added to the campaign. However, based on the chemical intuition of a chemist, one more experiment (e.g. DMAc, KOAc, BrettPhos, 120, 0.153) was performed. This additional experiment was never recommended before, but all the conditions were included in the searchspace at the start. Would there be a way to add this measurement during the course of the campaign, even though it was never recommended?

The solution that I see is to initiate a separate campaign from which one would extract a dataframe containing all the possible combinations with `all_combinations = campaign.recommend(batch_quantity = NUM_OF_POSSIBLE_COMBINATIONS)`. Then, using this dataframe, one could locate the entry (row) that corresponds to the additional experiment and extract it by the index. One would then add the measured target and add that specific measurement to the campaign with `campaign.add_measurements(additional_experiment)`. The full code would look something like this:

```
#all_combinations is a dataframe containing all possible experiments that BayBE can propose, obtained from a separate campaign
ADDITIONAL_EXPERIMENT_NO = 275 #Let's say that the custom experiment corresponds to entry 275 of all_combinations dataframe

custom_experiment = all_combinations.iloc[[ADDITIONAL_EXPERIMENT_NO]]
custom_experiment["Yield"] = 25 #for example
campaign.add_measurements(custom_experiment)
```

Is there maybe a more elegant way to achieve this? Also, bearing in mind your code, do you see any issues with such an approach, will it somehow impede with further recommendations? I also use a similar approach when I want to initiate a new campaign and add the previous measurements to it (with a new objective, for example). Some more efficient way to achieve this would be highly appreciated.

Cheers and thanks!

**Comment by Scienfitz** [comment, marked as answer]
Hi @ustojiljkoff ,

the solution is even simpler. 

As long as your additional point (or several points) are part of the searchspace, you can simply add them via `add_measurements`, no need for index matching, column sorting or any matching, baybe takes care of that.
Eg:
```python
my_custom_data = pd.DataFrame.from_records([
    {"Solvent": "DMAc", 
     "Base": "KOAc", 
     "Ligand": "BrettPhos", 
     "Temperature": 120, 
     "Concentration": 0.153, 
     "Yield": 42.0}
])

my_campaign.add_measurements(my_custom_data)
```

In fact you can do this any time, even before you trigger your first recommendation.


On another note: you can also get all possible combinations in case you have a purely discrete searchspace via accessing `your_campaign.searchspace.discrete.exp_rep`, although we advise not to meddle with these internal objects - in any case, I don't think its needed for your problem

**Reply by ustojiljkoff** [reply]
Hi @Scienfitz,

Many thanks for a prompt answer. Glad to see that there is such an easy and simple solution to my issue! I greatly appreciate your help!

**Reply by AdrianSosic** [reply]
Hi @ustojiljkoff. Glad that the problem could be solved so easily 👍🏼  Note that we are currently working intensively on the docs. The possibility to add existing data is already mentioned very briefly in the [Campaign User Guide](https://emdgroup.github.io/baybe/userguide/campaigns.html#adding-measurements) but we'll make this more clear in the next iteration and I'll soon also add an FAQ entry. 

Feel free an let us know if other things are unclear or if you need guidance somewhere else – this will surely help us improve the docs and the public API of the classes. Until then, happy coding 👨🏼‍💻

**Comment by Scienfitz** [comment]
closing as answered
